src/Symfony/Component/Validator/Constraints/NotSuspiciousValidator.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+use Symfony\Contracts\Translation\LocaleAwareInterface;
+
+/**
+ * @author Mathieu Lechat <mathieu.lechat@les-tilleuls.coop>
+ */
+class NotSuspiciousValidator extends ConstraintValidator implements LocaleAwareInterface
+{
+    private string $locale;
+
+    public function validate(mixed $value, Constraint $constraint)
+    {
+        if (!$constraint instanceof NotSuspicious) {
+            throw new UnexpectedTypeException($constraint, NotSuspicious::class);
+        }
+
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        if (!\is_scalar($value) && !$value instanceof \Stringable) {
+            throw new UnexpectedValueException($value, 'string');
+        }
+
+        $value = (string) $value;
+        if ('' === $value) {
+            return;
+        }
+
+        $checker = new \Spoofchecker();
+
+        $checker->setRestrictionLevel($constraint->restrictionLevel);
+
+        $allowedLocales = $constraint->profileLocales;
+        if ($constraint->addCurrentLocaleToProfile) {
+            $allowedLocales[] = $this->locale;
+        }
+        $checker->setAllowedLocales(implode(',', $allowedLocales));
+
+        $checker->setChecks($constraint->checks);
+
+        if (!$checker->isSuspicious($value)) {
+            return;
+        }
+
+        $this->context->buildViolation($constraint->message)
```

Comment:
can we compute a more specific message by checking rules one buy one until we find the one that made the check fail?

Reply:
I guess we could, but I’m not sure the added logic and cost (of calling `Spoofchecker::isSuspicious` for every check) would be worth it :thinking: 

Maybe when `&$errorCode` will work?

Reply:
I think giving precise hints to end users about why something fails is always worth it yes

Reply:
Does it make sense in the case of a spoof attempt though?

Even if we’re talking about a false positive I’m not sure this would help.

Reply:
I think it would yes. Legit users might be affected and we need to let them now why if possible.

Reply:
I wouldn’t be the one writing error messages then: how do you make the end user understand that he hid a combining mark in the previous one, or that we don’t want its input to be covered by Cyrillic, Greek or Cherokee? :sweat_smile:

Reply:
can you still add the logic? we'll propose something during the review

Reply:
Okay so a different error message per check. Do you think we also need to provide a different message for each restriction level?